src/metabase/driver/sql/query_processor.clj

```diff
@@ -979,136 +862,82 @@
-     :source-fields     (into {} (for [[source {:keys [this-level-fields]}] recursive
-                                       field                                this-level-fields]
-                                   field))}))
 
 (defn- apply-clauses
   "Like `apply-top-level-clauses`, but handles `source-query` as well, which needs to be handled in a special way
   because it is aliased."
   [driver honeysql-form {:keys [source-query], :as inner-query}]
-  (binding [*query*          inner-query
-            *source-aliases* (:source-fields (source-aliases driver inner-query))]
+  (binding [*query* inner-query]
     (if source-query
-      (apply-clauses-with-aliased-source-query-table
+      (apply-top-level-clauses
        driver
        (apply-source-query driver honeysql-form inner-query)
-       inner-query)
+       (dissoc inner-query :source-query))
       (apply-top-level-clauses driver honeysql-form inner-query))))
 
-(defn- expressions->subselect
-  [driver query]
-  (let [subselect (-> query
-                      (select-keys [:joins :source-table :source-query :source-metadata :expressions])
-                      (assoc :fields (-> (mbql.u/match (dissoc query :source-query :joins :expressions)
-                                           ;; remove the bucketing/binning operations from the source query -- we'll
-                                           ;; do that at the parent level
-                                           [:field id-or-name opts]
-                                           [:field id-or-name (dissoc opts :temporal-unit :binning)]
-                                           :expression
-                                           &match)
-                                         distinct)))]
-    (-> (mbql.u/replace query
-          [:expression expression-name]
-          [:field (escape-alias driver expression-name) {:base-type (:base_type (annotate/infer-expression-type &match))}]
-          ;; the outer select should not cast as the cast happens in the inner select
-          [:field (field-id :guard int?) field-info]
-          [:field field-id (assoc field-info ::outer-select true)])
-        (dissoc :source-table :joins :expressions :source-metadata)
-        (assoc :source-query subselect))))
-
-(defn- preprocess-query
-  [driver {:keys [expressions] :as query}]
-  (cond->> query
-    expressions (expressions->subselect driver)))
+(defmulti preprocess
+  "Do miscellaneous transformations to the MBQL before compiling the query. These changes are idempotent, so it is safe
+  to use this function in your own implementations of [[driver/mbql->native]], if you want to apply changes to the
+  same version of the query that we will ultimately be compiling."
+  {:arglists '([driver inner-query]), :added "0.42.0"}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
+
+(defmethod preprocess :sql
+  [_driver inner-query]
+  (nest-query/nest-expressions (add/add-alias-info inner-query)))
```

Comment:
We seem to call `add/add-alias-info` in `nest-expressions`? It is idempotent but why call twice?

Reply:
Nest expressions uses that info to figure out how to do the nesting, then once it does it just calls it again to recalculate everything because the old info no longer matches the query (isn't correct anymore) after the transformation